contrib/src/compression.rs

```diff
@@ -0,0 +1,169 @@
+use rocket::fairing::{Fairing, Info, Kind};
+
+#[cfg(feature = "brotli_compression")]
+use brotli;
+#[cfg(feature = "brotli_compression")]
+use brotli::enc::backward_references::BrotliEncoderMode;
+
+#[cfg(feature = "gzip_compression")]
+use flate2;
+#[cfg(feature = "gzip_compression")]
+use flate2::write::GzEncoder;
+
+#[derive(Debug, Default)]
+pub struct Compression;
+
+impl Compression {
+    /// This function creates a Compression to be used in your Rocket
+    /// instance. Add ```.attach(rocket_contrib::Compression::fairing())```
+    /// to your Rocket instance to use this fairing.
+    ///
+    /// # Returns
+    ///
+    /// A Compression instance.
+    pub fn fairing() -> Compression {
+        Compression::default()
+    }
+}
+
+impl Fairing for Compression {
+    fn info(&self) -> Info {
+        Info {
+            name: "Brotli and gzip compressors for responses",
+            kind: Kind::Response,
+        }
+    }
+
+    fn on_response(&self, request: &Request, response: &mut Response) {
+        let accept_headers: Vec<&str> = request.headers().get("Accept-Encoding").collect();
+        let mut content_header = false;
+        let mut content_header_br = false;
+        let mut content_header_gzip = false;
+        #[allow(unused_mut)]
+        let mut brotli_compressed = false;
+        #[allow(unused_mut)]
+        let mut gzip_compressed = false;
+        let mut image = false;
+        let headers = response.headers().clone();
+
+        if headers.contains("Content-Encoding") {
+            content_header = true;
+            if headers.get("Content-Encoding").any(|x| x == "br") {
+                content_header_br = true;
+            }
+            if headers.get("Content-Encoding").any(|x| x == "gzip") {
+                content_header_gzip = true;
+            }
+        }
+        if headers.get("Content-Type").any(|x| x.contains("image/")) {
+            image = true;
+        }
+        // The compression is done if the request supports brotli, the response
+        // does not have any Content-Encoding header or the Content-Encoding is
+        // brotli and the Content-Type is not an image (images compression
+        // ratio is minimum)
+        if cfg!(feature = "brotli_compression") && accept_headers.iter().any(|x| x.contains("br"))
+            && (!content_header || content_header_br) && !image
+        {
+            #[cfg(feature = "brotli_compression")]
+            {
+                brotli_compressed = true;
+                response.adjoin_header(Header::new("Content-Encoding", "br"));
+                let body = response.body_bytes();
+                if let Some(body) = body {
+                    let mut plain = Cursor::new(body);
+                    let mut compressed = Cursor::new(Vec::<u8>::new());
+                    let mut params = brotli::enc::BrotliEncoderInitParams();
+                    params.quality = 2;
+                    let content_type = headers
```

Comment:
What is this trying to do? You can get the raw content type string with `request.headers().get_one("Content-Type")`. But I don't think you need it. Just use `request.content_type()`.

Reply:
Here I collect and join in case the request had several `Content-Type` headers and some of them were `text/*`. Once more, this corner case maybe could be omited.